src/main/jbake/content/blog/2017/01/16/hawkular-metrics-with-dropwizard.adoc

```diff
@@ -0,0 +1,339 @@
+Every metric reported through this reporter will be tagged with the hostname.
+
+==== Filtering
+
+If you use Grafana with this sample app, you've probably noticed how annoying it is to find and select the metric you want to display,
+because it's flooded among tons of other metrics. And obviously, the more you store metrics, the more resources will be consumed. So you can filter out metrics you don't want.
+
+There's two kind of filters:
+
+- the usual built-in Dropwizard filters, that you can set using `HawkularReporterBuilder.filter` and by implementing link:http://metrics.dropwizard.io/3.1.0/apidocs/com/codahale/metrics/MetricFilter.html[MetricFilter]
+- another kind of filter that is very specific to the Hawkular reporter, called _MetricComposition_ and for which I must provide some details:
+
+As stated before, _Dropwizard_ has several metric types (gauges, meters, timers etc.), some of them being composed of multiple values. So they don't match 1-1 with _Hawkular_ metric types, which are made of simple values (basically, _doubles_ for _gauges_ and _longs_ for _counters_ -- there are other types but unused in the dropwizard reporter).
+
+In order not to loose any piece of data, _Dropwizard_ metrics are _exploded_ into several metrics in _Hawkular_. For instance, a Meter named _guava.cache.read_ will be translated into 4 gauges (_guava.cache.read.1minrt_, _guava.cache.read.5minrt_, _guava.cache.read.15minrt_, _guava.cache.read.meanrt_) and 1 counter (_guava.cache.read.count_) in _Hawkular_. The full translation table is link:https://github.com/hawkular/hawkular-dropwizard-reporter#how-it-works[described here].
+
+From the Dropwizard point of view, there is no metric called "guava.cache.read.1minrt". So you cannot filter it out with Dropwizard filters. However you can act on the "metric composition" in the Hawkular reporter. Either by providing the full metric name:
+
+``` java
+    // builder.
+      .setMetricComposition("guava.cache.read", Lists.newArrayList("1minrt", "meanrt", "count"))
+```
+
+or using regexp, as I'm doing in the sample app:
+
+``` java
+    HawkularReporter hawkularReporter = HawkularReporter.builder(registry, "com.hawkular.sample")
+            .addRegexTag(Pattern.compile(GuavaBackend.NAME + "\\..*"), "impl", GuavaBackend.NAME)
+            .addRegexTag(Pattern.compile(EhcacheBackend.NAME + "\\..*"), "impl", EhcacheBackend.NAME)
+            .addGlobalTag("hostname", hostname)
+            .prefixedWith(hostname + ".")
+            .setRegexMetricComposition(Pattern.compile("net\\.sf\\.ehcache"), Lists.newArrayList("mean", "meanrt", "5minrt", "98perc", "count"))
+            .setRegexMetricComposition(Pattern.compile(".*"), Lists.newArrayList("mean", "meanrt", "count"))
+            .build();
+```
+
+Here, we configure all _net.sf.ehcache.*_ metrics (EhCache middleware metrics) to provide their _mean_, _meanrt_, _5minrt_, _98perc_ and _count_ attributes. All other attributes will be discarded.
```

Comment:
No, it's "rate". I hope it's not too confusing... it's described in the doc here https://github.com/hawkular/hawkular-dropwizard-reporter#how-it-works